store.go

```diff
@@ -1534,124 +1680,210 @@ func (s *store) Mount(id, mountLabel string) (string, error) {
+		if modified, err := rlstore.Modified(); modified || err != nil {
+			rlstore.Load()
+		}
+		if rlstore.Exists(id) {
+			return rlstore.Unmount(id)
+		}
+	}
+	return ErrNotALayer
 }
 
 func (s *store) Changes(from, to string) ([]archive.Change, error) {
 	rlstore, err := s.LayerStore()
 	if err != nil {
 		return nil, err
 	}
-
-	rlstore.Lock()
-	defer rlstore.Unlock()
-	if modified, err := rlstore.Modified(); modified || err != nil {
-		rlstore.Load()
+	stores, err := s.ROLayerStores()
+	if err != nil {
+		return nil, err
 	}
-
-	return rlstore.Changes(from, to)
+	stores = append([]LayerStore{rlstore}, stores...)
+	for _, rlstore := range stores {
+		rlstore.Lock()
+		defer rlstore.Unlock()
+		if modified, err := rlstore.Modified(); modified || err != nil {
+			rlstore.Load()
+		}
+		if rlstore.Exists(to) {
+			return rlstore.Changes(from, to)
+		}
+	}
+	return nil, ErrNotALayer
 }
 
 func (s *store) DiffSize(from, to string) (int64, error) {
 	rlstore, err := s.LayerStore()
 	if err != nil {
 		return -1, err
 	}
-
-	rlstore.Lock()
-	defer rlstore.Unlock()
-	if modified, err := rlstore.Modified(); modified || err != nil {
-		rlstore.Load()
+	stores, err := s.ROLayerStores()
+	if err != nil {
+		return -1, err
 	}
-
-	return rlstore.DiffSize(from, to)
+	stores = append([]LayerStore{rlstore}, stores...)
+	for _, rlstore := range stores {
+		rlstore.Lock()
+		defer rlstore.Unlock()
+		if modified, err := rlstore.Modified(); modified || err != nil {
+			rlstore.Load()
+		}
+		if rlstore.Exists(to) {
+			return rlstore.DiffSize(from, to)
+		}
+	}
+	return -1, ErrNotALayer
 }
 
 func (s *store) Diff(from, to string) (io.ReadCloser, error) {
+
 	rlstore, err := s.LayerStore()
 	if err != nil {
 		return nil, err
 	}
-
-	rlstore.Lock()
-	defer rlstore.Unlock()
-	if modified, err := rlstore.Modified(); modified || err != nil {
-		rlstore.Load()
+	stores, err := s.ROLayerStores()
+	if err != nil {
+		return nil, err
 	}
-
-	return rlstore.Diff(from, to)
+	stores = append([]LayerStore{rlstore}, stores...)
+	for _, rlstore := range stores {
+		rlstore.Lock()
+		defer rlstore.Unlock()
+		if modified, err := rlstore.Modified(); modified || err != nil {
+			rlstore.Load()
+		}
+		if rlstore.Exists(to) {
+			return rlstore.Diff(from, to)
+		}
+	}
+	return nil, ErrNotALayer
 }
 
 func (s *store) ApplyDiff(to string, diff archive.Reader) (int64, error) {
 	rlstore, err := s.LayerStore()
 	if err != nil {
 		return -1, err
 	}
-
-	rlstore.Lock()
-	defer rlstore.Unlock()
-	if modified, err := rlstore.Modified(); modified || err != nil {
-		rlstore.Load()
+	stores, err := s.ROLayerStores()
+	if err != nil {
+		return -1, err
 	}
-
-	return rlstore.ApplyDiff(to, diff)
+	stores = append([]LayerStore{rlstore}, stores...)
+	for _, rlstore := range stores {
+		rlstore.Lock()
+		defer rlstore.Unlock()
+		if modified, err := rlstore.Modified(); modified || err != nil {
+			rlstore.Load()
+		}
+		if rlstore.Exists(to) {
+			return rlstore.ApplyDiff(to, diff)
+		}
+	}
```

Comment:
If I'm reading things right, if the layer to which the diff is being applied is in a read-only location, the overlay driver will create a layer with the same ID in the local read-write storage path (per the overlay driver's `getDiffPath()` method), but we won't have an entry for it in the read-write layerStore.